### PR TITLE
SONAR-8079 drop cache of rule repositories

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1319_create_table_rule_repositories.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+
+#
+# SonarQube 6.1
+#
+class CreateTableRuleRepositories < ActiveRecord::Migration
+
+  def self.up
+    create_table 'rule_repositories', :id => false do |t|
```
Comment (julienlancelot) [inline server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1319_create_table_rule_repositories.rb]: Not done in Java :) ?

File: server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1319_create_table_rule_repositories.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      t.column 'kee', :string, :limit => 200, :null => false
+      t.column 'language', :string, :limit => 20, :null => false
+      t.column 'name', :string, :limit => 4000, :null => false
```
Comment (julienlancelot) [inline server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1319_create_table_rule_repositories.rb]: Why 4000 ? In RULES.PLUGIN_NAME is limited to 255

File: server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1319_create_table_rule_repositories.rb (hunk 1 above)
Comment (simonbrandhof) [inline server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1319_create_table_rule_repositories.rb]: not yet, waiting for another PR !

File: server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1319_create_table_rule_repositories.rb (hunk 2 above)
Comment (simonbrandhof) [inline server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1319_create_table_rule_repositories.rb]: PLUGIN_NAME is..... repository key. There's no constraint with repository name in `RulesDefinition` as it was not persisted.

File: sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDao.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  public Optional<RuleRepositoryDto> selectByKey(DbSession dbSession, String key) {
+    return Optional.ofNullable(dbSession.getMapper(RuleRepositoryMapper.class).selectByKey(key));
+  }
+
+  public void truncate(DbSession dbSession) {
+    dbSession.getMapper(RuleRepositoryMapper.class).truncate();
+  }
+
+  public void insert(DbSession dbSession, Collection<RuleRepositoryDto> dtos) {
```
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDao.java]: Maybe would be best to use varargs (here or in another method), it's more easily readable in ut.

File: sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  }
+
+  public RuleRepositoryDto(String kee, String language, String name) {
+    this.kee = kee;
+    this.language = language;
+    this.name = name;
+  }
+
+  public String getKee() {
```
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java]: I would remove this getter

File: sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+  public String getLanguage() {
+    return language;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public RuleRepositoryDto setKee(String s) {
```
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java]: I think mybatis is doing introspection on member, not on setter, so I would rename it to setKey

File: sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    this.kee = s;
+    return this;
+  }
+
+  public RuleRepositoryDto setLanguage(String s) {
+    this.language = s;
```
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java]: In a perfect world, if every fields are mandatory, then a checkForNull would be great

File: sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+ */
+package org.sonar.db.rule;
+
+public class RuleRepositoryDto {
+
+  private String kee;
+  private String language;
+  private String name;
+
+  public RuleRepositoryDto() {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1186)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* CRITICAL 1 critical

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java [hunk 8]
```diff
@@ -114,6 +118,16 @@ public void start() {
     }
   }
 
+  private void persistRepositories(DbSession dbSession, List<RulesDefinition.Repository> repositories) {
+    dbClient.ruleRepositoryDao().truncate(dbSession);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java]: Do we really want to remove them all at each startup ?
Could be done later, but maybe we should only add new repos and remove no more existing ones.

File: server/sonar-server/src/test/java/org/sonar/server/rule/RegisterRulesTest.java [hunk 9]
```diff
@@ -168,7 +169,11 @@ public void update_and_remove_rules_on_changes() {
     assertThat(rule3).isNotNull();
     assertThat(rule3.getStatus()).isEqualTo(RuleStatus.READY);
 
+    // verify index
     assertThat(ruleIndex.search(new RuleQuery(), new SearchOptions()).getIds()).containsOnly(RULE_KEY1, RULE_KEY3);
+
+    // verify repositories
+    assertThat(dbClient.ruleRepositoryDao().selectAll(dbTester.getSession())).extracting(RuleRepositoryDto::getKee).containsOnly("fake");
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/rule/RegisterRulesTest.java]: You should add some repositories at the begin of the test in order to verify that they have been removed

File: server/sonar-server/src/test/java/org/sonar/server/rule/ws/RepositoriesActionTest.java [hunk 10]
```diff
@@ -19,67 +19,52 @@
...
 
     newRequest().execute().assertJson(this.getClass(), "repositories.json");
     newRequest().setParam("language", "xoo").execute().assertJson(this.getClass(), "repositories_xoo.json");
     newRequest().setParam("language", "ws").execute().assertJson(this.getClass(), "repositories_ws.json");
     newRequest().setParam("q", "common").execute().assertJson(this.getClass(), "repositories_common.json");
     newRequest().setParam("q", "squid").execute().assertJson(this.getClass(), "repositories_squid.json");
     newRequest().setParam("q", "sonar").execute().assertJson(this.getClass(), "repositories_sonar.json");
-    newRequest().setParam("q", "sonar").setParam("ps", "2").execute().assertJson(this.getClass(), "repositories_limited.json");
     newRequest().setParam("ps", "4").execute().assertJson(this.getClass(), "repositories.json");
     newRequest().setParam("ps", "100").execute().assertJson(this.getClass(), "repositories.json");
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/rule/ws/RepositoriesActionTest.java]: These 2 lines should be removed

**Comment by julienlancelot** [comment]
Minor feedback to apply and LGTM !

File: sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java (hunk 6 above)
Comment (simonbrandhof) [inline sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java]: no, it has been decided to not put @CheckForNull and @Nullable on non-null columns.

File: server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java (hunk 8 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java]: IMO this is over-complex compared to the number of rows involved here.

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java]: Ok, I trust you on perf subject !

File: server/sonar-server/src/test/java/org/sonar/server/rule/ws/RepositoriesActionTest.java (hunk 10 above)
Comment (simonbrandhof) [inline server/sonar-server/src/test/java/org/sonar/server/rule/ws/RepositoriesActionTest.java]: good catch

File: server/sonar-server/src/test/java/org/sonar/server/rule/RegisterRulesTest.java (hunk 9 above)
Comment (simonbrandhof) [inline server/sonar-server/src/test/java/org/sonar/server/rule/RegisterRulesTest.java]: you're right. I'm doing that.

File: sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java (hunk 6 above)
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/rule/RuleRepositoryDto.java]: Sorry, I was thinking about Preconditions.checkForNull, not the annotation

**Pull request merged**